server/src/main/java/io/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,327 @@
+
+    // hostAndPort -> DiscoveryDruidNode
+    private final Map<String, DiscoveryDruidNode> nodes = new ConcurrentHashMap<>();
+
+    private final PathChildrenCache cache;
+    private final ExecutorService cacheExecutor;
+
+    private final ExecutorService listenerExecutor;
+
+    private final List<DruidNodeDiscovery.Listener> nodeListeners = new ArrayList();
+
+    NodeTypeWatcher(
+        ExecutorService listenerExecutor,
+        CuratorFramework curatorFramework,
+        String basePath,
+        ObjectMapper jsonMapper,
+        String nodeType
+    )
+    {
+      this.listenerExecutor = listenerExecutor;
+      this.curatorFramework = curatorFramework;
+      this.nodeType = nodeType;
+      this.jsonMapper = jsonMapper;
+
+      // This is required to be single threaded from Docs in PathChildrenCache;
+      this.cacheExecutor = Execs.singleThreaded(String.format("NodeTypeWatcher[%s]-[%%d]", nodeType));
+      this.cache = new PathChildrenCache(
+          curatorFramework,
+          ZKPaths.makePath(basePath, nodeType),
+          true,
+          true,
+          cacheExecutor
+      );
+    }
+
+    @Override
+    public Set<DiscoveryDruidNode> getAllNodes()
+    {
+      return new ImmutableSet.Builder().addAll(nodes.values()).build();
+    }
+
+    @Override
+    public synchronized void registerListener(DruidNodeDiscovery.Listener listener)
+    {
+      for (DiscoveryDruidNode node : nodes.values()) {
+        listenerExecutor.submit(() -> {
+          try {
+            listener.nodeAdded(node);
+          }
+          catch (Exception ex) {
+            log.error(ex, "Exception occured in DiscoveryDruidNode.nodeAdded(node=[%s]) in listener [%s].", node, listener);
+          }
+        });
+      }
+
+      nodeListeners.add(listener);
+    }
+
+    @Override
+    public synchronized void childEvent(CuratorFramework client, PathChildrenCacheEvent event) throws Exception
+    {
+      try {
+        switch (event.getType()) {
+          case CHILD_ADDED: {
+            final byte[] data;
+            try {
+              data = curatorFramework.getData().decompressed().forPath(event.getData().getPath());
+            }
+            catch (Exception ex) {
+              log.error(ex, "Failed to get data for path [%s]. Ignoring event [%s].", event.getData().getPath(), event.getType());
+              return;
+            }
+
+            DiscoveryDruidNode druidNode = jsonMapper.readValue(
+                data,
+                DiscoveryDruidNode.class
+            );
+
+            if (!nodeType.equals(druidNode.getNodeType())) {
+              log.error(
+                  "Node[%s:%s] add is discovered by node watcher of nodeType [%s]. Ignored.",
+                  druidNode.getNodeType(),
+                  druidNode,
+                  nodeType
+              );
+              return;
+            }
+
+            log.info("Received event [%s] for Node[%s:%s].", event.getType(), druidNode.getNodeType(), druidNode);
```

Comment:
This log appears to be repeated amongst the various parts of the switch/case, maybe move it up above the switch case just so that we are certain this stuff gets logged?